shard.go

```diff
@@ -36,23 +36,64 @@ func (s *cacheShard) get(key string, hashedKey uint64) ([]byte, error) {
 	}
 
 	wrappedEntry, err := s.entries.Get(int(itemIndex))
+	s.lock.RUnlock()
 	if err != nil {
-		s.lock.RUnlock()
 		s.miss()
 		return nil, err
 	}
-	if entryKey := readKeyFromEntry(wrappedEntry); key != entryKey {
-		if s.isVerbose {
-			s.logger.Printf("Collision detected. Both %q and %q have the same hash %x", key, entryKey, hashedKey)
+
+	return wrappedEntry, err
+}
+
+func (s *cacheShard) getWithInfo(key string, hashedKey uint64) (entry []byte, resp Response, err error) {
+	currentTime := uint64(s.clock.epoch())
+	wrappedEntry, err := s.getWrappedEntry(key, hashedKey)
+	if err == nil {
+		s.lock.RLock()
+		if entryKey := readKeyFromEntry(wrappedEntry); key != entryKey {
+			if s.isVerbose {
+				s.logger.Printf("Collision detected. Both %q and %q have the same hash %x", key, entryKey, hashedKey)
+			}
+			s.lock.RUnlock()
+			s.collision()
+			return entry, resp, ErrEntryNotFound
+		}
+
+		oldestTimeStamp := readTimestampFromEntry(wrappedEntry)
+		if currentTime-oldestTimeStamp >= s.lifeWindow {
+			s.lock.RUnlock()
+			// @TODO: when Expired is non-default value return err as nil as the resp will have proper entry status
+			resp.EntryStatus = Expired
+			return entry, resp, ErrEntryIsDead
+		}
+		entry := readEntry(wrappedEntry)
+		s.lock.RUnlock()
+		s.hit()
+		return entry, resp, nil
+	}
+	// it is nil & error
+	return wrappedEntry, resp, err
+}
+
+func (s *cacheShard) get(key string, hashedKey uint64) ([]byte, error) {
```

Comment:
Can you move this up before `getWrappedEntry`. Just for readability it's easier when helper functions are after caller.

Reply:
done